proposals/KEEP-0389-kdoc-streamline-ambiguous-KDoc-links.md

```diff
@@ -0,0 +1,1752 @@
+import something.* // Explicit star importing scope
+
+class MyClass {
+    // Type scope
+
+    companion object {
+        // Static type scope
+        fun staticFoo() {}
+    }
+
+    fun <T> foo() {
+        // Local scope
+    }
+}
+```
+
+Then it should iterate through all the scopes 
+and gather declarations with the given name, according to the following priorities:
+1. Class
+2. Function
+3. Property
+
+If there are any declarations of the same kind found in the current scope,
+all these declarations should be immediately returned.
+
+#### Multi-segment names treatment
+
+The handling of multi-segment names here is a bit trickier.
+A multi-segment name can be either a relative name or a global FQN.
+And there is no way to tell which one we are looking at before the resolution is performed.
+Such links are first considered as relative links,
+and then they are treated as global links.
+
+1. [Relative name resolution](#relative-name-resolution)
+2. [Global name resolution](#global-name-resolution)
+
+Here it's crucial to understand the following algorithm that helps us to
+transform multi-segment name search into a single-segment name search:
+
+##### Scope reduction algorithm
+
+This algorithm will be actively used in the following sections.
+It helps us to apply the same logic as we used for short names to the multi-segment name search.
+
+Imagine that we have some relative name, e.g. `A.B.C` in the following case:
+```kotlin
+class Global {
+    class A {
+        class B {
+            class C {
+                companion object {
+                    /**
+                     * [A.B.C]
+                     */
+                    fun foo() {}
+                }
+            }
+        }
+    }
+}
+```
+
+A relative name is a multi-segment name that is relative to some position visible from the current context.
+In this case, the relative name is `A.B.C` and it's relative to the `Global` class.
+
+The resolver has already calculated the list of visible scopes 
+when searching for short names in [Short names treatment](#short-names-treatment).
+If a relative link is resolvable, then it must be relative to any of these visible scopes.
+
+To reduce this multi-segment search to a single-segment one, 
+all these scopes have to be slightly modified.
+
+`A.B.C` implies that somewhere there is a chain of two nested classifiers `A` and `B` with `B`
+containing some member `C`.
+The resolver should try to find a classifier `A` in each of these scopes and, if found,
+replace the original scope with the member scope of this class.
+If no suitable classes are found in some scope, this scope should be discarded.
+Then the resolver should take all the modified scopes and repeat the search for nested classifiers (`B` in this case)
+until just the short name `C` is left.
+Then it's just a single-segment name search in the constructed list of scopes.
+
+Note that currently it's prohibited to have a variable as a link segment.
```

Comment:
Not only a variable, but also a function? Typealias?
`non-last/final link segment` instead of `a link segment`?